### PR TITLE
feat: Add support for spot instances node groups.

### PR DESCRIPTION
**Pull request by psoares**
# PR o'clock

## Description

Please explain the changes you made here and link to any relevant issues.
This is to implement feature request #1107.
It allows creating nodegroups with spot instances, as described in https://aws.amazon.com/about-aws/whats-new/2020/12/amazon-eks-support-ec2-spot-instances-managed-node-groups/

example usage:
```
module "eks" {
  // [...]
  node_groups = {
    spot-inst-managed = {
      name             = "spot-managed-${var.cluster_name}"
      desired_capacity = 2
      max_capacity     = 20
      min_capacity     = 2
      instance_type    = ["m5.large", "m5a.large", "m5d.large", "m5ad.large"]
      capacity_type    = "SPOT"
    }
```

### Checklist

- [x] CI tests are passing
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation

### CODE REVIEW

**Comment by zmingxie** [comment]
Could you also update the example here with the new `capacity_type` option? 
https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/examples/launch_templates_with_managed_node_groups/main.tf

This will be easier for future references

**Comment by brannondorsey** [comment]
Managed spot instance node pools add support for mixed instance types. Would it be possible to make the `instance_type` accept a list of instance types in addition to a string?

**Comment by psoares** [comment]
@brannondorsey I'm updating the documentation to make it obvious.

**Comment by brannondorsey** [comment]
Thanks for that docs change 😸

We're currently using this module and have had to make our managed spot instance node groups out of band without Terraform. I appreciate this small PR which will make our lives much easier. 🙏

**Comment by brannondorsey** [comment]
Any idea when this might land in a new release? We're looking to use this functionality this week and are contemplating if we should just vendor this branch into our code base or wait for a new release of the module with these changes included.

**Comment by psoares** [comment]
What's the process to get a PR reviewed here?

**Comment by worldofgeese** [comment]
> What's the process to get a PR reviewed here?

Looks like @barryib has write access. Pinging him for you.

**Comment by psoares** [comment]
> > What's the process to get a PR reviewed here?
> 
> Looks like @barryib has write access. Pinging him for you.
Merci @barryib!

**Comment by worldofgeese** [comment]
Maybe if I try @dpiddockcmp also

**Comment by duboisf** [comment]
Hi @max-rocket-internet! Would it be possible to get a review for this PR pretty please?

**Comment by sumitg892** [comment]
When will this PR get merged?

**Comment by psoares** [comment]
> When will this PR get merged?

This is my first PR... I have no experience on how long it usually takes to get a review.

**Review by huiser — APPROVED** [review]

**Comment by sumitg892** [comment]
@psoares, have you please tested your fix that it works?

File: modules/node_groups/node_groups.tf
Comment (duboisf) [inline modules/node_groups/node_groups.tf:18]: What was the reason that the check was changed from verifying if the `launch_template_id` is `""` instead of `null`? I forked and tried this PR and the resulting `instance_types` was a `t3.large`. I didn't have a `launch_template_id` anywhere (`workers_group_defaults`, `node_group_defaults` or `node_groups`).

Changing the predicate back to `each.value["launch_template_id"] != null` resulted in the `instance_types` being picked up from the `instance_type` set in `var.node_groups`.

Comment (psoares) [reply modules/node_groups/node_groups.tf:18]: Thanks @duboisf. I don't recall exactly how, but at least on terraform 0.14, null doesn't work for me, whereas "" does.
What's your version of Terraform? Should I be testing in 0.12 instead?

Comment (psoares) [reply modules/node_groups/node_groups.tf:18]: It might be a side effect of https://github.com/terraform-aws-modules/terraform-aws-eks/commit/571da60aea10034cef20c435c2b49933ab5ede47

Comment (psoares) [reply modules/node_groups/node_groups.tf]: Ok, I understand now. It's indeed related to the PR I mention above. I was doing my tests by editing files directly in my .terraform folder and forgot to update the value of launch_template_id to null in local.tf. I reverted that value back to null and pushed a new commit. Please check now.

**Review by duboisf — APPROVED** [review]
Tested it :+1:

File: modules/node_groups/node_groups.tf
Comment (duboisf) [reply modules/node_groups/node_groups.tf]: Tested it, it works :+1:

**Comment by psoares** [comment]
> Tested it 👍

Thank you 😉
Any idea why the merge is still blocked?

**Comment by duboisf** [comment]
Yes we need one of the maintainers to approve the PR

**Review by mknapcok — APPROVED** [review]

**Comment by antimack** [comment]
also interested in this, any estimations?

**Comment by darkhelmet** [comment]
How will this work with a launch template and wanting to pick multiple instance types? Do I create a launch template per instance type and then a node group per launch template?

Basically, I have a launch template, and I want a spot instance node group of "any of these instance types" (which it looks like your example supports, if you don't specify a launch template).

Maybe that's a separate thing? I can leave the instance type off the launch template, but then the node group module ignores the instance type I specify on the group itself. Should that be the case?

**Comment by psoares** [comment]
> How will this work with a launch template and wanting to pick multiple instance types? Do I create a launch template per instance type and then a node group per launch template?
> 
> Basically, I have a launch template, and I want a spot instance node group of "any of these instance types" (which it looks like your example supports, if you don't specify a launch template).
> 
> Maybe that's a separate thing? I can leave the instance type off the launch template, but then the node group module ignores the instance type I specify on the group itself. Should that be the case?
> 
> Edit: Looking at the AWS help when creating a new node group I see `When using a launch template do not specify instanceTypes, diskSize or remoteAccess in the node group configuration. These properties must be configured within the launch template itself.` which suggests it's doing things properly.

I've been scratching my head on this one for a few days. I even opened #1192. I'll let you know if I find a way. So far I've been able to do a launch template, but not override it's parameters with a launch configuration to handle multiple instance types...

**Comment by barryib** [comment]
I just merged https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1129 and it adds the SPOT support for managed node groups.

@psoares can we close this PR as it duplicates https://github.com/terraform-aws-modules/terraform-aws-eks/pull/1129 ?

**Comment by psoares** [comment]
Closing since the feature was implemented and merged through #1129.

**Comment by psoares** [comment]
@barryib, should we redo this PR to fix #1211 ?

**Comment by barryib** [comment]
> @barryib, should we reopen this PR to fix #1211 ?

Nope. Thanks and sorry for the delay.

**Pull request closed without merging**
